osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/service/ProtocolServiceLookup.java

```diff
@@ -0,0 +1,31 @@
+package org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.configuration.service;
+
+import java.util.List;
+
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.Protocol;
+import org.opensmartgridplatform.adapter.protocol.dlms.exceptions.ProtocolAdapterException;
+
+public class ProtocolServiceLookup {
+
+    private final List<ProtocolService> protocolServices;
+
+    public ProtocolServiceLookup(final List<ProtocolService> protocolServices) {
+        this.protocolServices = protocolServices;
+    }
+
+    public GetConfigurationObjectService lookupGetService(final Protocol protocol) throws ProtocolAdapterException {
+        return this.lookupProtocolService(GetConfigurationObjectService.class, protocol);
+    }
+
+    public SetConfigurationObjectService lookupSetService(final Protocol protocol) throws ProtocolAdapterException {
+        return this.lookupProtocolService(SetConfigurationObjectService.class, protocol);
+    }
+
```

Comment:
this is a nice concept, but with Spring you can just grab the bean from the applicationContext by @Component(name=GetConfigurationObjectServiceSmr5).

In that case this class is not needed, but in the caller you can say:

@Autowired
private ApplicationContext сontext;

context.getBean("GetConfigurationObjectService"+protcolName/Version.) and cast to the generic GetConfigurationObjectService

Reply:
The logic for handling a protocol (see `boolean handles(Protocol)` implementations) is not as straightforward as you suggest. For instance we have SMR 5.0 and 5.1 which should both be handled. Therefore I prefer a single dedicated lookup class than trying to jump through (less understandable) bean lookup hoops with Spring. The latter would also mean this knowledge about which bean names and protocol suffixes are needed gets spread around the code, which is a less maintainable. 

Finally, I like implementation classes themselves to specify which cases they can handle, rather than recording this in some remote location (be it the lookup class, or Spring bean annotations). In this way the lookup class only needs to know that there are a bunch of ProtocolServices which implement Get/Set functionality, nothing about specific protocols or implementation class names.